### ISSUE TITLE
Trigger mousemove events while scrolling

### ISSUE THREAD

**Issue by 1j01** [hollowdoor/dom_autoscroller#14]
Objects being dragged generally only update on `mousemove`, and thus lag behind when autoscrolling, unless the object is isolated from the scroll offset of the container (as apparently is the case in dragula).
For example if using jquery.nestable as in [this jsfiddle](http://jsfiddle.net/1j01/p41d3y9a/1/), the dragged element doesn't update while scrolling unless you continuously move the mouse.

dom-autoscroller should trigger `mousemove` events while scrolling to make these applications work seamlessly without additional code. It'll have to avoid retriggering itself from the events (which should be fairly easy), and should probably avoid sending extra `mousemove` events when the mouse is already moving, but hopefully in a way that keeps it smooth (i.e. only avoid triggering the event if a `mousemove` was received since the last animation frame, rather than since some fixed amount of time)

**Comment by hollowdoor** [comment]
Use version 2.1.3, or greater. ;)

http://jsfiddle.net/hollowdoor/rby97c46/1/

Please leave this issue open for a while so people that have semver blocking update to 2.1.3 will be notified.

**Comment by 1j01** [comment]
I did, or at least I tried to. JSFiddle apparently isn't letting me remove external resources.
[Here's a new fiddle, still demonstrating the issue,](https://jsfiddle.net/1j01/7h9necx3/) that is definitely using 2.1.3

**Comment by hollowdoor** [comment]
I've looked at it closer. 2.1.3 works better, but the dragged element moves away from the pointer when scrolling then suddenly catches up when the pointer is moved again.

Also what the hell? I tried to add a custom mouse event to the fiddle, the screen went gray, and a dialog that I can't dismiss appeared!

**Comment by 1j01** [comment]
Yeah, that's what I was meaning to describe. (I don't know why I said "lags" behind, it just stays behind.)

**Comment by hollowdoor** [comment]
Well it does lag. By not following the pointer.

LOL. Someone tried to add auto scrolling to nestable in 2013.

https://github.com/dbushell/Nestable/pull/46

For posterity the related issue for nestable https://github.com/dbushell/Nestable/issues/17

**Comment by hollowdoor** [comment]
Someone figured out how to do nested with dragula too. https://github.com/bevacqua/dragula/issues/31

I'm liking the idea of calling this a bug with nestable. If only that repo owner was more active.

I would prefer not firing a mouse event. That could mess with future compatibilities big time.

**Comment by hollowdoor** [comment]
Also by the way. If you want to have simulated mouse events trigger you can do that in the `autoScroll` method option of dom-autoscroller. `autoScroll` is called every mouse move regardless of scroll happening. You can return false after triggering to stop the extra scroll.

**Edit**: actually I might be wrong about `autoScroll` method being a viable option. It causes too much recursion, doesn't scroll, or scrolls badly for everything I've tried.

**Comment by 1j01** [comment]
That's the exact time when you don't need to send a `mousemove` event; it would need to be sent during the animation, but ideally not duplicated when there was already a `mousemove` event.

**Comment by hollowdoor** [comment]
I know. I tried to delay it with a timeout, but it didn't quite work. It caused staggering because it was out of sync.

**Comment by hollowdoor** [comment]
Alright. I think I just figured out something.

The object that controls nestable can be retrieved with jQuery.prototype.data.

``` javascript
var dd = $('.dd').nestable({ /* config options */ });
console.log(dd.data('nestable')) //Prints the nestable object
```

How strange. Anyway. `nestable` exposes it's methods from `jQuery.prototype.data('nestable')`.

I'll add a moving property to the dom-autoscroller instance for tracking that. The scrolling property got neglected in the process of making version 2 so I need to expose that again.

moving (or pointerMoving), and scrolling can be used to track the state of the scroller for when some other lib needs updating.

Sadly dragging for nestable still depends on a mouse event so an event might still have to be created if not also triggered. There is a `dragElem` property on the nestable object so that is a possibility if it can be retrieved at the right time.

I really hate to put a custom event inside dom-autoscroller. :(

That will be my last resort. I think I might go the way of adapters. Especially for compatibility for older libraries that might be entirely abandoned one day, or change hands, and implementations. I don't want to insert code into the body of dom-autoscroller to support some library that might not be here tomorrow. If more than one library has the same problem I'll consider adding the code to dom-autoscroller directly.

**Comment by 1j01** [comment]
You're considering adding custom code to support a specific library? Over triggering an event so that all libraries past and future work seamlessly without having to know about dom-autoscroller or visa-versa?
What's the problem with triggering a `mousemove` event? (a "custom event" is a different thing btw)

**Comment by hollowdoor** [comment]
It doesn't look like it would take a simple mousemove event too me. I can make a mousemove then trigger it. If I do that then I have to make it's x/y change for some other library that needs that update. Remember we're talking about a mouse being still is the reason we would want to simulate. And the x/y movement to be reflected comes from the scrolling.

To me this doesn't look like a simple matter of "all libraries past and future work seamlessly". I can see the utility, but I also see problems arising where two external needs will conflict.

I'm not saying that it's "custom code to support a specific library". I'm saying I don't see what you're asking for as being anything other than that. I don't have a collection of libraries for testing this one hypothetical scenario of all libraries that consume x/y. Libraries that need x/y changing on still pointer, libraries that need x/y to act normal on still pointer, or libraries that need no mousemove event when the pointer is still. I can create these libraries for the purpose of testing?

Now there's this problem of keep mouse moving on some condition even when it's not moving.

For dom-autoscroller the `autoScroll` method is one of those things to take care of situations like this -- where external dependencies might conflict. `autoScroll` gives control to a user to stop scrolling during some condition.

I'm considering the idea. Keep convincing me. Sometimes I need a while to think. To you it might appear like there's one simple way to implement. To me for some concepts it's never one simple way. It can be multiple hard ways, or multiple simple ways. Like the other issue for speed on margin/edge proximity that you proposed. You convinced me of one simple way by showing the solution. I just haven't seen the way for this particular issue yet. Call it my limitation, or whatever you want.

**Comment by hollowdoor** [comment]
I need your opinion.

We know that a simulated event can't be done inside the mousemove listener so it would have to be in the animation call.

The question is should the simulated mousemove be triggered before scroll tick, or after scroll tick?

I still don't know if I want to do it, but these details will help me decide.

This is my birthday weekend. I might check up on things, but I probably won't be talking to you until next Tuesday. Have a good weekend.

**Comment by 1j01** [comment]
It works best if it happens before the scroll (i.e. outside the `setTimeout`), that makes it nice and smooth.

Have a good one! 😄

**Comment by hollowdoor** [comment]
After some research I have found several situations where the object can't be removed from it's context, or there isn't an object (for instance highlighting) making scrolling not work correctly as in the library nestable. Though for dragging -- an element can have it's own context some pointer actions are bound to the current context. Context being the parent element.

I think my hesitance is due to the nature of this library being the only one that does this auto scrolling behavior exclusively. I've put myself in a strange position where I can't see the forest for the trees even though I intend to scroll the forest. :)

~~There is only one problem I have which is touch. There is no way to determine absolutely if a screen is a touch screen so the emulation will probably have to be mouse only. Until I can figure out a way to make touch, and mouse emulation work I think the emulation behavior should have an option to turn it off.~~ As browsers that support touch events do fire touch, and mouse for all of those types of interactions (except hover) it might work to just fire both. It would be consistent behavior.

There is also the possibility that some apps will not need/want mouse movement on scroll. So an option covers that case. I have yet to decide if it should be a default false, or true option because I don't know how many container types would use the emulation behavior. I'm erring towards true, but whatever we choose it better be for the long term. Some testing will reveal a narrow set of cases where the emulation is desired.

I have a couple things I want to get done before I start on this so leave any feedback you can before I begin work on the emulation. I would appreciate any criticism/advice you can provide. I might begin work on mouse emulation by the end of the week, or some time next week.

**Comment by hollowdoor** [comment]
My computer went on the fritz today. Three crashes in a row. I'll be trying to start this next week if I'm not prevented by my wishy washy pc. I have several ideas that I want to test. If I'm going to do this I don't want to get stuck with a strange system that isn't consistent. I'll put the emulation in, and warn people that it's experimental in the readme. I'll include an option to turn it off. In version 3 I'll remove the option if no one complains.

I really wanted to start on this. As I also really really wanted to update the readme, but I'm afraid of losing work, or data in a crash.

**Comment by 1j01** [comment]
Maybe you should try [Cloud9](https://c9.io). I use it for whatever I can, it's really nice.

**Comment by hollowdoor** [comment]
Looks interesting. I just got a new computer so I'm not so worried about that anymore. :)

Back to business as usual.

**Comment by hollowdoor** [comment]
@1j01 I've added a mousemove dispatch to dom-autoscroller on my own machine. It appears to be working.

I set it to update client(X/Y), and page(X/Y) so far. I tested using a javascript dotted mouse trail using 12 dots. The mouse trail follows the pointer just like it would normally. The code I grabbed for the mouse trail had it's point updater set to window. I changed it to the sub element of dom-autoscroller for testing.

The results are somewhat satisfying, but the movement of scrolling is a bit jittery while dragging around a mouse trail. I tried to optimize the mousemove dispatcher to mitigate jitteryness, but some mild jitter might be acceptable. I might try different timing for the dispatch (after scroll update maybe) to see if the jitter can be reduced.

I've added the option `sync` (I might change the name) to turn the dispatcher on, and off. Though if someone turns it off they will still get all the code for the dispatcher even if they aren't using it. So that's something to consider. Either way the option can be there so people can opt out if they aren't satisfied with the result.

There are other things I'm not sure of. One being other event properties. Right now I have the two types of coordinate properties set. I don't know what other properties might be required for other usages.

The special mousemove dispatcher library that I created for this purpose is here https://github.com/hollowdoor/dom_mousemove_dispatcher

I looked at other dispatch libraries, but pretty much all of them are general so their entire code base might be imported (rollup does tree shaking) leading to inflating dom-autoscroller being most dispatch libraries handle every edge case possible for compatibility reasons.

My dispatcher is incomplete as I am not able to test in all environments to make sure the dispatcher works cross browser. So I put some trust in code I found on the internet to add some old browser support for now.

Some more testing is required before I push/publish, and after I push/publish. I don't know if I'm leaving any details out, but that's the update on this issue for now.

I'm still trying to think of a better way to do this without changing dom-autoscroller very much, but I'm not sure of other options that would work. I'm thinking of adding a tick listener method that can be called on the scroll update to see where that takes things.

There is also a problem of extra dispatches on external libraries. There might be some side effects worth worrying about for when a library gets it's mousemove updates when it's not supposed to.
